includes/promoted-jobs/class-wp-job-manager-promoted-jobs-notifications.php

```diff
@@ -0,0 +1,335 @@
+				continue;
+			}
+			if ( 'job_title' === $field ) {
+				$meta_fields[ $field ] = $post->post_title;
+				continue;
+			}
+			$meta_fields[ $field ] = get_post_meta( $post_id, '_' . $field, true );
+		}
+		return $meta_fields;
+	}
+
+	/**
+	 * Map values to the data we are watching for changes.
+	 *
+	 * @param array $values Values of the job fields.
+	 */
+	private function map_to_watched_fields( $values ) {
+		return [
+			'job_title'       => $values['job']['job_title'],
+			'job_description' => $values['job']['job_description'],
+			'company_name'    => $values['company']['company_name'],
+			'application'     => $values['job']['application'],
+			'job_location'    => $values['job']['job_location'],
+		];
+	}
+
+	/**
+	 * Extract required job listing meta from the $_POST array.
+	 *
+	 * @param WP_Post $post Post object.
+	 */
+	private function get_post_data( $post ) {
+		$post_data = [];
+		foreach ( $this->watched_fields as $field ) {
+			if ( 'job_description' === $field ) {
+				$post_data[ $field ] = $post->post_content;
+				continue;
+			}
+			if ( 'job_title' === $field ) {
+				$post_data[ $field ] = $post->post_title;
+				continue;
+			}
+			// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.NonceVerification.Missing -- Input sanitized in registered post meta config;
+			$post_data[ $field ] = isset( $_POST[ "_$field" ] ) ? wp_unslash( $_POST[ "_$field" ] ) : '';
+		}
+		return $post_data;
+	}
+
+	/**
+	 * Determine if any job field has changed.
+	 *
+	 * @param array $current_values Current values of the job.
+	 * @param array $new_values New values of the job.
+	 */
+	private function post_has_changed( $current_values, $new_values ) {
+		foreach ( $current_values as $key => $value ) {
+			if ( $value !== $new_values[ $key ] ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Check if a job is promoted.
+	 *
+	 * @param int $post_id Post ID.
+	 * @return bool
+	 */
+	private function is_promoted_job( $post_id ) {
+		return get_post_meta( $post_id, '_promoted', true );
+	}
+
+	/**
+	 * Check if we should notify wpjobmanager.com of a change.
+	 *
+	 * @param int $post_id Post ID.
+	 * @return bool
+	 */
+	public function should_notify( $post_id ) {
+		if ( 'job_listing' !== get_post_type( $post_id ) ) {
+			return false;
+		}
+		if ( ! $this->is_promoted_job( $post_id ) ) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Notify wpjobmanager.com when a Job Listing is trashed.
+	 *
+	 * @param int $post_id Post ID.
+	 * @return void
+	 */
+	public function promoted_job_trashed( $post_id ) {
+		if ( ! $this->should_notify( $post_id ) ) {
+			return;
+		}
+		$this->send_notification();
+	}
+
+	/**
+	 * Notify wpjobmanager.com when a Job Listing is updated.
+	 *
+	 * @param int   $post_id Post ID.
+	 * @param array $values Values.
+	 * @return void
+	 */
+	public function promoted_job_updated( $post_id, $values ) {
```

Comment:
I'm wondering if control it through the `save_post` for both options would be better to control the change in a single method. Also, it could catch the job listing being updated through other ways too (some other code calling `wp_update_post`, for example) which I think these 2 methods wouldn't catch, right?

Reply:
7ebc3e1f265fa7c6b7f4995c47483d896fcdb845
`post_updated` and `update_postmeta` hooks cover everything including `status`.

Reply:
Nice! Code looks great now!

I just noticed that after changing post and post meta (tested in WP Admin), it's calling the notification 2 times. It seems one is coming from an ajax request and the other not.
Do you have any idea how to trigger it a single time only after both were changed?

Reply:
No idea but since we now have only `_promoted` as the meta we are interested in and it doesn't get updated in admin, it won't call the notification twice anymore.

Reply:
Oh! Good point!
I was a little worried about how complicated would be this. That was easy! 🙌 haha